src/web/lib/components/ThemeCustomBackgroundPicker/index.js

```diff
@@ -0,0 +1,367 @@
+import {
+  CUSTOM_BACKGROUND_MAXIMUM_SIZE,
+  CUSTOM_BACKGROUND_ALLOWED_TYPES
+} from "../../../../lib/constants";
+
+import "./index.scss";
+
+import iconHAlignLeft from "./icon_align_left.svg";
+import iconVAlignCenter from "./icon_align_center.svg";
+
+import iconLoading from "./icon_loading.svg";
+import iconInfo from "../../../../images/info-16.svg";
+
+export class ThemeCustomBackgroundPicker extends React.Component {
+  constructor(props) {
+    super(props);
+  }
+
+  handleImageAdd = ({ name }) => {
+    this.props.addCustomBackground({
+      name,
+      tiling: "no-repeat",
+      alignment: "left top"
+    });
+  };
+
+  // HACK: Prevent default on drag to stop text highlighting
+  // https://github.com/clauderic/react-sortable-hoc/issues/253#issuecomment-350223723
+  handleSortStart = (_, event) => {
+    event.preventDefault();
+  };
+
+  handleSortEnd = ({ oldIndex, newIndex }) => {
+    this.props.moveCustomBackground({ oldIndex, newIndex });
+  };
+
+  render() {
+    const { addImage, updateImage } = this.props;
+    return (
+      <form className="custom-background" onSubmit={e => e.preventDefault()}>
+        <label htmlFor="customBackground">
+          <h2>Custom background image</h2>
+        </label>
+
+        <ImageImporter
+          {...{ addImage, updateImage, onImport: this.handleImageAdd }}
+        >
+          {({ importing, errors, ImportButton }) => (
+            <div className="add-image">
+              <ImportButton label="Upload image" />
+              {importing && (
+                <img className="status-icon importing" src={iconLoading} />
+              )}
+              {errors && (
+                <React.Fragment>
+                  <img
+                    className="status-icon errors"
+                    src={iconInfo}
+                    title={JSON.stringify(errors)}
+                  />
+                  <Modal>
+                    <ul className="errors">
+                      {errors.tooLarge && (
+                        <li>The image is too large. (1MB maximum size)</li>
+                      )}
+                      {errors.wrongType && (
+                        <li>The file is not an accepted image type.</li>
+                      )}
+                    </ul>
+                  </Modal>
+                </React.Fragment>
+              )}
+            </div>
+          )}
+        </ImageImporter>
+
+        <BackgroundList
+          {...this.props}
+          helperClass="dragHelper"
+          useDragHandle={true}
+          onSortStart={this.handleSortStart}
+          onSortEnd={this.handleSortEnd}
+        />
+      </form>
+    );
+  }
+}
+
+const modalRoot = document.getElementById("modal");
+
+class Modal extends React.Component {
+  constructor(props) {
+    super(props);
+    this.el = document.createElement("div");
+  }
+
+  componentDidMount() {
+    modalRoot.appendChild(this.el);
+    setTimeout(() => this.el.classList.add("fade-in"), 10);
```

Comment:
This feels kind of dirty, but couldn't think of a better way to give these dialogs a transition in. Could probably also use a timed transition out, or maybe self-dismiss on any click